renderlib/gesture/gesture.h

```diff
@@ -0,0 +1,388 @@
+      if (mbIndex >= kButtonsCount) {
+        return;
+      }
+
+      reset(mbs[mbIndex]);
+    }
+
+    void reset()
+    {
+      for (int mbIndex = 0; mbIndex < kButtonsCount; ++mbIndex) {
+        reset(mbs[mbIndex]);
+      }
+    }
+
+    // Call this function at the end of a frame before polling new events.
+    void consume()
+    {
+      // Any button release event that we didn't explicitly use must be consumed, and the button
+      // state must be reset.
+      for (int mbIndex = 0; mbIndex < kButtonsCount; ++mbIndex) {
+        if (mbs[mbIndex].action == kRelease) {
+          reset(mbs[mbIndex]);
+        }
+      }
+    }
+
+    struct Button
+    {
+      Action action = kNone;
+      // A bitwise combination of Mods keys (such as alt, ctrl, shift, etc...)
+      int modifier = 0;
+      // The screen coordinates the button was initially pressed, zero if no action
+      glm::vec2 pressedPosition = {};
+      // The drag vector from pressedPosition, this is still valid if pressed is kRelease and
+      // it can be used to differentiate between a click action to be execute on button release
+      // and the end of a click-drag gesture.
+      glm::vec2 drag = {};
+      // 0: unconstrained, 1: horizontal, 2: vertical
+      DragConstrain dragConstraint = kUnconstrained;
+      bool doubleClick = false;
+      // The last time the button was pressed. This is used to detect double-click events.
+      double triggerTime = 0;
+    };
+
+    Button mbs[kButtonsCount];
+    glm::vec2 cursorPos = glm::vec2(0);
+
+    bool hasButtonAction(ButtonId id, int mods)
+    {
+      return mbs[id].action != kNone && (mods == 0 || ((mbs[id].modifier & mods) > 0));
+    }
+
+    // Reset is typically executed by the command or tool that consumes the button release action.
+    static void reset(Button& button)
+    {
+      button.action = kNone;
+      button.modifier = 0;
+      button.pressedPosition = glm::vec2(0);
+      button.drag = glm::vec2(0);
+      button.dragConstraint = kUnconstrained;
+      button.doubleClick = false;
+      // Do not reset triggerTime here: we need to retain its value to detect double-clicks
+    }
+  };
+  Input input;
+
+  // A minimalist graphics system to draw interactive GUI elements on screen
+  struct Graphics
+  {
+    static constexpr int kInvalidVertexIndex = -1;
+
+    // We support multiple command lists. Each list will draw with different
+    // settings. The number of independent list is currently hardcoded but
+    // it could be made configurable if we would store an array of function
+    // pointers to setup each draw pass.
+    enum class CommandSequence : int
+    {
+      k3dDepthTested = 0, //< for viewport depth compositing 3d GUI elements
+      k3dStacked = 1,     //< for 3d GUI elements overlay
+      k2dScreen = 2       //< for topmost 2d elements (widgets and buttons)
+    };
+    static constexpr int kNumCommandsLists = 3;
+
+    struct Command
```

Comment:
Would be nice to have a comment here explaining what a Command is? I found it while looking through the blog but may not be immediately obvious for people touching this code for the first time.